src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -91,13 +102,164 @@ static void CreateHashDistributedTable(Oid relationId, char *distributionColumnN
 									   char *colocateWithTableName,
 									   int shardCount, int replicationFactor);
 static Oid ColumnType(Oid relationId, char *columnName);
+static void ErrorIfNotSuitableToGetSize(text *relationText);
+static void CopyLocalDataIntoShards(Oid relationId);
+static List * TupleDescColumnNameList(TupleDesc tupleDescriptor);
+static void ErrorIfNotSuitableToGetSize(text *relationText);
 static void CopyLocalDataIntoShards(Oid relationId);
 static List * TupleDescColumnNameList(TupleDesc tupleDescriptor);
 
 /* exports for SQL callable functions */
 PG_FUNCTION_INFO_V1(master_create_distributed_table);
 PG_FUNCTION_INFO_V1(create_distributed_table);
 PG_FUNCTION_INFO_V1(create_reference_table);
+PG_FUNCTION_INFO_V1(citus_table_size);
+PG_FUNCTION_INFO_V1(citus_total_relation_size);
+PG_FUNCTION_INFO_V1(citus_relation_size);
+
+/*
+ * SupportFunctionForCitusSize is a helper function for all citus size functions.
+ *
+ */
+static uint32
+SupportFunctionForCitusSize(text *relationText, CitusSizeFunctionTypes sizeFunctionType)
+{
+	Oid distributedRelationId = ResolveRelationId(relationText);
+	DistTableCacheEntry *distributedTableCacheEntry = NULL;
+
+	/* Build shard qualified name. */
+	Oid schemaId = get_rel_namespace(distributedRelationId);
+	char *schemaName = get_namespace_name(schemaId);
+
+	int shardIndex = 0;
+	uint32 totalRelationSize = 0;
+
+	MultiConnection *connection = NULL;
+
+	distributedTableCacheEntry = DistributedTableCacheEntry(distributedRelationId);
+
+	for (shardIndex = 0; shardIndex <
+		 distributedTableCacheEntry->shardIntervalArrayLength; shardIndex++)
+	{
+		ShardPlacement *placement =
+			distributedTableCacheEntry->arrayOfPlacementArrays[shardIndex];
+		int64 shardId = placement->shardId;
+
+		connection = GetNodeConnection(0, placement->nodeName, placement->nodePort); /* TODO : I'm not sure about the flag ! */
+
+		char *shardName = get_rel_name(distributedRelationId);
+		AppendShardIdToName(&shardName, shardId);
+
+		char *shardQualifiedName = quote_qualified_identifier(schemaName, shardName);
+		char *quotedShardName = quote_literal_cstr(shardQualifiedName);
+
+		StringInfo tableSizeQuery = makeStringInfo();
+		PGresult *queryResult = NULL;
+
+		bool cstoreTable = CStoreTable(distributedRelationId);
+		if (cstoreTable)
+		{
+			appendStringInfo(tableSizeQuery, SHARD_CSTORE_TABLE_SIZE_QUERY,
+							 quotedShardName);
+		}
+		else
+		{
+			if (sizeFunctionType == CITUS_TOTAL_RELATION_SIZE)
+			{
+				appendStringInfo(tableSizeQuery, SHARD_TOTAL_RELATION_SIZE_QUERY,
+								 quotedShardName);
+			}
+			else if (sizeFunctionType == CITUS_TABLE_SIZE)
+			{
+				appendStringInfo(tableSizeQuery, SHARD_TABLE_SIZE_QUERY, quotedShardName);
+			}
+			else if (sizeFunctionType == CITUS_RELATION_SIZE)
+			{
+				appendStringInfo(tableSizeQuery, SHARD_RELATION_SIZE_QUERY,
+								 quotedShardName);
+			}
+		}
+
+		int executeCommand = ExecuteOptionalRemoteCommand(connection,
```

Comment:
Instead of running a command for every shard, can we construct one big SELECT statement per worker such that we only make a single round-trip per worker?